tokio/src/io/util/read_to_end.rs

```diff
@@ -68,19 +68,29 @@ fn poll_read_to_end<V: VecU8, R: AsyncRead + ?Sized>(
     // of data to return. Simply tacking on an extra DEFAULT_BUF_SIZE space every
     // time is 4,500 times (!) slower than this if the reader has a very small
     // amount of data to return.
-    buf.reserve(32);
+    let try_small_read = buf.reserve(32);
 
     // Get a ReadBuf into the vector.
     let mut read_buf = buf.get_read_buf();
 
+    let poll_result;
     let filled_before = read_buf.filled().len();
-    let poll_result = read.poll_read(cx, &mut read_buf);
-    let filled_after = read_buf.filled().len();
-    let n = filled_after - filled_before;
-
+    let filled_after;
+    if try_small_read {
+        let mut small_buf = Vec::with_capacity(32);
+        let mut small_read_buf = ReadBuf::new(&mut small_buf);
+        poll_result = read.poll_read(cx, &mut small_read_buf);
+        let filled = small_read_buf.filled().len();
+        read_buf.put_slice(&small_buf[..filled]);
+        filled_after = filled_before + filled;
+    } else {
+        poll_result = read.poll_read(cx, &mut read_buf);
+        filled_after = read_buf.filled().len();
+    };
```

Comment:
style nit, take it or leave it: i think this simplifies to
```suggestion
    let filled_after = if try_small_read {
        let mut small_buf = Vec::with_capacity(32);
        let mut small_read_buf = ReadBuf::new(&mut small_buf);
        poll_result = read.poll_read(cx, &mut small_read_buf);
        let filled = small_read_buf.filled().len();
        read_buf.put_slice(&small_buf[..filled]);
        filled_before + filled
    } else {
        poll_result = read.poll_read(cx, &mut read_buf);
        read_buf.filled().len()
    };
```